File: app/src/main/java/com/pandatone/kumiwake/kumiwake/QuickMode.java
package com.pandatone.kumiwake.kumiwake;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.graphics.Point;
import android.graphics.Rect;
import android.os.Build;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.AppCompatEditText;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.transition.Slide;
import android.view.View;
import android.view.ViewTreeObserver;
import android.view.Window;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.ImageView;
import android.widget.SeekBar;
import android.widget.TextView;

import com.pandatone.kumiwake.R;

import java.util.ArrayList;

import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * Created by atsushi_2 on 2016/05/02.
 */
public class QuickMode extends AppCompatActivity implements TextWatcher {
    AppCompatEditText GpNoEditText, MbNoEditText;
    @SuppressLint("StaticFieldLeak")
    static TextView errorGroup, errorMember, manNoText, womanNoText;
    ImageView backgroundImage;
    SeekBar seekBar;
    CheckBox even_fm_ratio_check, even_person_ratio_check;
    static int memberNo, manNo, womanNo;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            getWindow().requestFeature(Window.FEATURE_CONTENT_TRANSITIONS);
            getWindow().setExitTransition(new Slide());
        }
        setContentView(R.layout.quick_mode);
        ButterKnife.bind(this);
        findViews();
        even_fm_ratio_check.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                if (even_fm_ratio_check.isChecked() == true) {
                    even_person_ratio_check.setEnabled(true);
                } else {
                    even_person_ratio_check.setChecked(false);
                    even_person_ratio_check.setEnabled(false);
                }
            }
        });
        seekBar.setEnabled(false);
        Point size = new Point();
        getWindowManager().getDefaultDisplay().getSize(size);
        int screenHeight = size.y;
        backgroundImage.getLayoutParams().height=screenHeight;
        MbNoEditText.addTextChangedListener(this);
    }

    public void findViews(){
        GpNoEditText = (AppCompatEditText) findViewById(R.id.input_group_no);
        MbNoEditText = (AppCompatEditText) findViewById(R.id.input_member_no);
        errorGroup = (TextView) findViewById(R.id.error_group_no);
        errorMember = (TextView) findViewById(R.id.error_member_no);
        manNoText = (TextView) findViewById(R.id.man_number);
        womanNoText = (TextView) findViewById(R.id.woman_number);
        seekBar = (SeekBar) findViewById(R.id.sex_seek_bar);
        even_fm_ratio_check = (CheckBox) findViewById(R.id.even_fm_ratio_check);
        even_person_ratio_check = (CheckBox) findViewById(R.id.even_person_ratio_check);
        backgroundImage =(ImageView)findViewById(R.id.background_img);
    }

    @Override
    public void beforeTextChanged(CharSequence s, int start, int count, int after) {
    }

    @Override
    public void onTextChanged(CharSequence s, int start, int before, int count) {
    }

    @Override
    public void afterTextChanged(Editable s) {
        if (!s.toString().equals("")) {
            memberNo = Integer.parseInt(s.toString());
            seekBar.setEnabled(true);
        } else {
            memberNo = 0;
            seekBar.setEnabled(false);
        }
        seekBar.setMax(memberNo);
        seekBar.setProgress(memberNo);

        seekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromTouch) {
                manNo = progress;
                womanNo = memberNo - progress;
                manNoText.setText(String.valueOf(manNo));
                womanNoText.setText(String.valueOf(womanNo));
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
            }
        });
    }

    @OnClick(R.id.quick_kumiwake_button)
    void onClicked() {
        String group_no = GpNoEditText.getText().toString();
        String member_no = MbNoEditText.getText().toString();

        if (TextUtils.isEmpty(group_no)) {
            errorGroup.setText(R.string.error_empty_group_no);
        }
        if (TextUtils.isEmpty(member_no)) {
            errorMember.setText(R.string.error_empty_member_no);
        } else if (TextUtils.isEmpty(group_no)) {
            errorGroup.setText(R.string.error_empty_group_no);
        } else if (Integer.parseInt(group_no) > memberNo) {
            errorGroup.setText(R.string.number_of_groups_is_much_too);
            errorMember.setText("");
        } else {
            ArrayList<String> womanList = new ArrayList<String>();
            ArrayList<String> manList = CreateManList(manNo, womanNo);
            ArrayList<String> groupList = new ArrayList<String>();
            for (int i = 1; i <= womanNo; i++) {
                womanList.add(getText(R.string.member) + "♡" + String.valueOf(i));
            }
            for (int i = 1; i <= Integer.parseInt(group_no); i++) {
                groupList.add(getText(R.string.group) + " " + String.valueOf(i));
            }
            Intent intent = new Intent(this, QuickKumiwakeConfirmation.class);
            intent.putStringArrayListExtra("QuickModeManList", manList);
            intent.putStringArrayListExtra("QuickModeWomanList", womanList);
            intent.putStringArrayListExtra("QuickModeGroupList", groupList);
            intent.putExtra("EvenFMRatio", even_fm_ratio_check.isChecked());
            intent.putExtra("EvenPersonRatio", even_person_ratio_check.isChecked());
            startActivity(intent);
            overridePendingTransition(R.anim.in_right, R.anim.out_left);
        }
    }

    public ArrayList<String> CreateManList(int manNo, int womanNo) {
        ArrayList<String> manList = new ArrayList<String>();
        if (womanNo == 0) {
            for (int i = 1; i <= manNo; i++) {
                manList.add(getText(R.string.member) + " " + String.valueOf(i));
            }
        } else {
            for (int i = 1; i <= manNo; i++) {
                manList.add(getText(R.string.member) + "♠" + String.valueOf(i));
            }
        }
        return manList;
    }
}

File: app/src/main/java/com/pandatone/kumiwake/kumiwake/QuickMode.kt
package com.pandatone.kumiwake.kumiwake

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.transition.Slide
import android.view.MotionEvent
import android.view.View
import android.view.Window
import android.view.inputmethod.InputMethodManager
import android.widget.Button
import android.widget.EditText
import android.widget.SeekBar
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.content.ContextCompat
import androidx.core.view.GestureDetectorCompat
import com.google.firebase.analytics.FirebaseAnalytics
import com.pandatone.kumiwake.FirebaseAnalyticsEvents
import com.pandatone.kumiwake.KumiwakeArrayKeys
import com.pandatone.kumiwake.KumiwakeCustomKeys
import com.pandatone.kumiwake.ModeKeys
import com.pandatone.kumiwake.MyGestureListener
import com.pandatone.kumiwake.PublicMethods
import com.pandatone.kumiwake.R
import com.pandatone.kumiwake.StatusHolder
import com.pandatone.kumiwake.Theme
import com.pandatone.kumiwake.databinding.QuickModeBinding
import com.pandatone.kumiwake.member.function.Member


/**
 * Created by atsushi_2 on 2016/05/02.
 */
class QuickMode : AppCompatActivity(), TextWatcher {
    private lateinit var binding: QuickModeBinding

    private var memberNo: Int = 0
    private var manNo: Int = 0
    private var womanNo: Int = 0

    private lateinit var mDetector: GestureDetectorCompat

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        window.requestFeature(Window.FEATURE_CONTENT_TRANSITIONS)
        window.exitTransition = Slide()
        FirebaseAnalyticsEvents.firebaseAnalytics = FirebaseAnalytics.getInstance(this)
        //テーマを設定
        if (StatusHolder.mode == ModeKeys.Sekigime.key) {
            PublicMethods.setStatus(this, Theme.Sekigime.primaryColor)
        } else {
            PublicMethods.setStatus(this, Theme.Kumiwake.primaryColor)
        }
        binding = QuickModeBinding.inflate(layoutInflater)
        setContentView(binding.root)
        if (StatusHolder.mode == ModeKeys.Sekigime.key) {
            val layout = findViewById<ConstraintLayout>(R.id.quick_layout)
            layout.background = ContextCompat.getDrawable(this, R.drawable.sekigime_background)
        }
        binding.sexSeekBar.isEnabled = false
        binding.memberNoForm.addTextChangedListener(this)
        findViewById<Button>(R.id.quick_kumiwake_btn).setOnClickListener { onNextClicked() }

        Toast.makeText(this, getText(R.string.double_tap), Toast.LENGTH_SHORT).show()
        val groupNoInput = findViewById<EditText>(R.id.group_no_form)
        val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        mDetector = GestureDetectorCompat(this, MyGestureListener(imm, groupNoInput))
        mDetector.setOnDoubleTapListener(MyGestureListener(imm, groupNoInput))
        setupSeekBar()
    }

    //スクロールビューの場合こっち呼ぶ
    override fun dispatchTouchEvent(event: MotionEvent): Boolean {
        super.dispatchTouchEvent(event)
        mDetector.onTouchEvent(event)
        return super.onTouchEvent(event)
    }

    override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}

    override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}

    override fun afterTextChanged(s: Editable) {
        if (s.toString() != "") {
            memberNo = Integer.parseInt(s.toString())
            binding.sexSeekBar.isEnabled = true
        } else {
            memberNo = 0
            binding.sexSeekBar.isEnabled = false
        }
        binding.sexSeekBar.max = memberNo
        binding.sexSeekBar.progress = memberNo
        manNo = memberNo
        womanNo = 0
        binding.manNumberTxt.text = manNo.toString()
        binding.womanNumberTxt.text = womanNo.toString()
    }

    private fun setupSeekBar() {
        binding.sexSeekBar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onStartTrackingTouch(sexSeekbar: SeekBar) {}

            override fun onProgressChanged(
                sexSeekbar: SeekBar,
                progress: Int,
                fromTouch: Boolean
            ) {
                manNo = progress
                womanNo = memberNo - manNo
                binding.manNumberTxt.text = manNo.toString()
                binding.womanNumberTxt.text = womanNo.toString()
            }

            override fun onStopTrackingTouch(sexSeekbar: SeekBar) {}
        })
    }

    private fun onNextClicked() {
        val groupNo = binding.groupNoForm.text.toString()
        val memNo = binding.memberNoForm.text.toString()
        val errorMemberNo = findViewById<TextView>(R.id.error_member_no_txt)
        val errorGroupNo = findViewById<TextView>(R.id.error_group_no_txt)
        errorMemberNo.visibility = View.GONE
        errorGroupNo.visibility = View.GONE
        errorMemberNo.text = ""
        errorGroupNo.text = ""

        when {
            TextUtils.isEmpty(memNo) -> {
                errorMemberNo.visibility = View.VISIBLE
                errorMemberNo.setText(R.string.error_empty_member_no)
            }

            TextUtils.isEmpty(groupNo) -> {
                errorGroupNo.visibility = View.VISIBLE
                errorGroupNo.setText(R.string.error_empty_group_no)
            }

            groupNo == "0" -> {
                errorGroupNo.visibility = View.VISIBLE
                errorGroupNo.setText(R.string.require_correct_No)
            }

            Integer.parseInt(groupNo) > memberNo -> {
                errorGroupNo.visibility = View.VISIBLE
                errorGroupNo.setText(R.string.number_of_groups_is_much_too)
            }

            else -> {
                val memberList = createMemberList(manNo, memberNo - manNo)
                val groupList = PublicMethods.initialGroupArray(
                    this,
                    Integer.parseInt(groupNo),
                    memberList.size,
                    false
                )
                val intent = Intent(this, KumiwakeConfirmation::class.java)
                intent.putExtra(KumiwakeArrayKeys.MEMBER_LIST.key, memberList)
                intent.putExtra(KumiwakeArrayKeys.GROUP_LIST.key, groupList)
                intent.putExtra(
                    KumiwakeCustomKeys.EVEN_FM_RATIO.key,
                    binding.evenFmRatioCheck.isChecked
                )
                startActivity(intent)
                overridePendingTransition(R.anim.in_right, R.anim.out_left)
                //Add Firebase
                if (StatusHolder.mode == ModeKeys.Sekigime.key) {
                    FirebaseAnalyticsEvents.countEvent(
                        memberNo,
                        Integer.parseInt(groupNo),
                        FirebaseAnalyticsEvents.FunctionKeys.SekigimeQuick.key
                    )
                } else {
                    FirebaseAnalyticsEvents.countEvent(
                        memberNo,
                        Integer.parseInt(groupNo),
                        FirebaseAnalyticsEvents.FunctionKeys.KumiwakeQuick.key
                    )
                }
            }
        }
    }

    private fun createMemberList(manNo: Int, womanNo: Int): ArrayList<Member> {
        val memberList = ArrayList<Member>()
        if (manNo == 0 || womanNo == 0) {
            for (i in 1..manNo + womanNo) {
                val planeMember = Member(
                    i,
                    getText(R.string.member).toString() + " " + i.toString(),
                    StatusHolder.none,
                    0,
                    "",
                    "",
                    -1
                )
                memberList.add(planeMember)
            }
        } else {
            for (i in 1..manNo) {
                val man = Member(
                    i,
                    getText(R.string.member).toString() + "♠" + i.toString(),
                    getString(R.string.man),
                    0,
                    "",
                    "",
                    -1
                )
                memberList.add(man)
            }
            for (i in 1..womanNo) {
                val woman = Member(
                    i + 1000,
                    getText(R.string.member).toString() + "♡" + i.toString(),
                    getString(R.string.woman),
                    0,
                    "",
                    "",
                    -1
                )
                memberList.add(woman)
            }
        }
        return memberList
    }
}
